flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/TableEnvironment.scala

```diff
@@ -502,26 +513,140 @@ abstract class TableEnvironment(val config: TableConfig) {
     *   tEnv.sql(s"SELECT * FROM $table")
     * }}}
     *
-    * @param query The SQL query to evaluate.
+    * @param sql The SQL string to evaluate.
     * @return The result of the query as Table.
     */
-  def sql(query: String): Table = {
+  @deprecated
+  def sql(sql: String): Table = {
     val planner = new FlinkPlannerImpl(getFrameworkConfig, getPlanner, getTypeFactory)
     // parse the sql query
-    val parsed = planner.parse(query)
+    val parsed = planner.parse(sql)
     // validate the sql query
     val validated = planner.validate(parsed)
     // transform to a relational tree
     val relational = planner.rel(validated)
-
     new Table(this, LogicalRelNode(relational.rel))
   }
 
   /**
+    * Evaluates a SQL Select query on registered tables and retrieves the result as a
+    * [[Table]].
+    *
+    * All tables referenced by the query must be registered in the TableEnvironment. But
+    * [[Table.toString]] will automatically register an unique table name and return the
+    * table name. So it allows to call SQL directly on tables like this:
+    *
+    * {{{
+    *   val table: Table = ...
+    *   // the table is not registered to the table environment
+    *   tEnv.sqlSelect(s"SELECT * FROM $table")
+    * }}}
+    *
+    * @param sql The SQL string to evaluate.
+    * @return The result of the query as Table or null of the DML insert operation.
+    */
+  def sqlQuery(sql: String): Table = {
+    val planner = new FlinkPlannerImpl(getFrameworkConfig, getPlanner, getTypeFactory)
+    // parse the sql query
+    val parsed = planner.parse(sql)
+    if (null != parsed && parsed.getKind.belongsTo(SqlKind.QUERY)) {
+      // validate the sql query
+      val validated = planner.validate(parsed)
+      // transform to a relational tree
+      val relational = planner.rel(validated)
+      new Table(this, LogicalRelNode(relational.rel))
+    } else {
+      throw new TableException(
+        "Unsupported sql query! sqlQuery Only accept SELECT, UNION, INTERSECT, EXCEPT, VALUES, " +
```

Comment:
SqlParser.parseStmt() actually call the SqlParser.parseQuery, so they're the same. Could not help us to distinguish the sql type, so use SqlKind here, SqlKind.QUERY consists of: SELECT, EXCEPT, INTERSECT, UNION, VALUES, ORDER_BY, EXPLICIT_TABLE.